### PR TITLE
Enable Ruby to dynamically link against a custom GC library

### PR DESCRIPTION
**Pull request by eileencodes**
Ticket: https://bugs.ruby-lang.org/issues/20351

We made it so Ruby can optionally build and link against a librubygc, so
that gc routines can be moved into a shared library.

Usage:

The PR adds two new configuration flags: enabling a shared gc and enabling shared gc that uses relative paths.

To dynamically link the GC object, configure Ruby using the following:

```c
./autogen.sh
./configure --enable-shared-gc
make -j
```

To dynamically link the GC object and load it on a relative path, configure Ruby using the following:

```c
./autogen.sh
./configure --enable-shared-gc --enable-load-relative-gc
make -j
```

To replace the GC add a custom GC do the following:

Make a custom GC at `/custom/rubygc.c` (the path and filename aren't significant, but the output filename is):

```c
#include <stdio.h>
#include <stdlib.h>

void
GC_Init(void) {
    fprintf(stderr, "===== >>> Loading our custom GC file\n");
}
```

Compile with:

Clang for macos:

```
clang -shared -dynamiclib custom/rubygc.c -o custom/librubygc.3.4.dylib
```

GCC for linux:

```
gcc -shared -fPIC custom/rubygc.c -o custom/librubygc.so.3.4.0
```

Load the custom GC:

Macos:

```
DYLD_LIBRARY_PATH=./custom ./ruby -e '1'
===== >>> Loading our custom GC file
```

Linux:

```
LD_PRELOAD=./custom/librubygc.so.3.4.0 ruby -e '1'
===== >>> Loading our custom GC file
```

It's important to note that to start we're only going to support loading a shared object for the GC on linux and macos. All other operating systems will statically link the GC instead.

The PR also adds CI builds for macos and linux that test all the combinations of `--enable-shared` and `--enable-shared-gc` with `--enable-load-relative` and `--enable-load-relative-gc` respectively.

### CODE REVIEW

File: template/Makefile.in [hunk 1]
```diff
@@ -175,6 +181,17 @@ TEENY=	@TEENY@
...
+LIBRUBYGC_A     = @LIBRUBYGC_A@
+LIBRUBYGC_SO    = @LIBRUBYGC_SO@
+LIBRUBYGC_SONAME= @LIBRUBYGC_SONAME@
+LIBRUBYGC_ALIASES= @LIBRUBYGC_ALIASES@
+LIBRUBYGC      = @LIBRUBYGC@
+LIBRUBYGCARG    = @LIBRUBYGCARG@
+LIBRUBYGCARG_STATIC = @LIBRUBYGCARG_STATIC@
+LIBRUBYGCARG_SHARED = @LIBRUBYGCARG_SHARED@
+LIBRUBYGC_A_OBJS = @LIBRUBYGC_A_OBJS@
+LIBRUBYGC_RELATIVE = @LIBRUBYGC_RELATIVE@
```
Comment (nobu) [inline template/Makefile.in:193]: What is this variable used for?

**Comment by nobu** [comment]
Is the change to win32/Makefile.sub needed?

**Comment by kateinoigakukun** [comment]
How about keeping the default GC system statically linked to simplify the build and reduce configuration variants? If I understand correctly, it's ok to place the default GC system code and a custom one in a same memory and symbol spaces.
In that case, we can just declare `GC_Init` as weak symbol and check it at runtime to select whether system should be inited:

```c
__attribute__((weak)) void GC_Init_hook(void);

void
ruby_init(void)
{
  if (GC_Init_hook) {
    GC_Init_hook()
  } else {
    GC_Init_default();
  }
```

**Comment by nobu** [comment]
I succeeded to load hooks dynamically on Linux, but weak symbols seems unsupported on macOS.

```diff
diff --git i/eval.c w/eval.c
index a405857f9ae..587b74ca5f9 100644
--- i/eval.c
+++ w/eval.c
@@ -93,9 +93,16 @@ ruby_setup(void)
     return state;
 }
 
+RBIMPL_DYNAMIC_HOOK
+void
+ruby_init_hook(void)
+{
+}
+
 void
 ruby_init(void)
 {
+    ruby_init_hook();
     int state = ruby_setup();
     if (state) {
         if (RTEST(ruby_debug)) {
diff --git i/include/ruby/internal/attr/weakref.h w/include/ruby/internal/attr/weakref.h
index f118bb62b86..852ac2950b9 100644
--- i/include/ruby/internal/attr/weakref.h
+++ w/include/ruby/internal/attr/weakref.h
@@ -29,4 +29,10 @@
 # define RBIMPL_ATTR_WEAKREF(sym) /* void */
 #endif
 
+#if RBIMPL_HAS_ATTRIBUTE(weak)
+# define RBIMPL_DYNAMIC_HOOK RUBY_FUNC_EXPORTED __attribute__((weak))
+#else
+# define RBIMPL_DYNAMIC_HOOK static inline
+#endif
+
 #endif /* RBIMPL_ATTR_WEAKREF_H */
diff --git i/ruby.c w/ruby.c
index 5c1d44ec6df..da3106ae461 100644
--- i/ruby.c
+++ w/ruby.c
@@ -1731,9 +1731,7 @@ void rb_call_builtin_inits(void);
 
 // Initialize extra optional exts linked statically.
 // This empty definition will be replaced with the actual strong symbol by linker.
-#if RBIMPL_HAS_ATTRIBUTE(weak)
-__attribute__((weak))
-#endif
+RBIMPL_DYNAMIC_HOOK
 void
 Init_extra_exts(void)
 {
```

With `ruby_init_hook.c`
```c
#include <stdio.h>
#include <stdlib.h>

void
ruby_init_hook(void)
{
    fputs("Loading from CRuby\n", stderr);
}
```

and `Init_extra_exts.c`.
```c
#include <stdio.h>
#include <stdlib.h>

void
Init_extra_exts(void)
{
    fputs("Init_extra_exts: loaded\n", stderr);
}
```

Then:
```sh-session
$ for x in init_extra_exts ruby_init_hook; do gcc -shared -o $x.so ../src/$x.c; done
$ make runnable
$ LD_PRELOAD=./init_extra_exts.so:./ruby_init_hook.so ./bin/ruby -ep
Loading from CRuby
Init_extra_exts: loaded
```

**Comment by eightbitraptor** [comment]
Thanks @nobu and @kateinoigakukun for the feedback and the help with `__attribute__((weak))`. It looks like macOS might support this with the attibute name `weak_import`? We're going to look into this further and do some benchmarking to see how this compares to our existing approach.

@nobu the changes to `win32/Makefile.sub` were necessary to get CI passing, because we introduce a new object. Even though we aren't explicitly targeting Windows for this feature Windows still needs to know how to compile and link the static archive in order to build Ruby correctly.

**Comment by kateinoigakukun** [comment]
>  It looks like macOS might support this with the attibute name weak_import?

Yes, it supports the semantics by combination of `weak_import` and `-undefined dynamic_lookup`. However, `-undefined dynamic_lookup` is already deprecated, it might be better to simply use `dlopen` instead of `DYLD_LIBRARY_PATH` to inject custom GC libraries.

**Comment by eightbitraptor** [comment]
> How about keeping the default GC system statically linked to simplify the build and reduce configuration variants? In other words, do we really need to introduce a new .so in Ruby build?

@kateinoigakukun @nobu While we investigate alternative ways to achieve our goals here, I was wondering if you could clarify what concerns you have with adding another (optional) `.so` to Ruby's build system. Are the concerns purely about making the build process more complex, or are there other issues as well?

**Comment by kateinoigakukun** [comment]
Yes, my main concern here is "purely" about complicating the build system.

- As you know, it is not easy to add a new `.so` component in our current build system and is very error-prune unfortunately.
- We don't need to expose the GC as library to Ruby users and it's not a goal of this feature while `libruby.so` is used by some applications to embed Ruby.
- Introducing a build variant requires twice the effort to cover all build combinations in CI and local development. I think having a unified build for this feature makes it easier to maintain for us all.

**Comment by eightbitraptor** [comment]
> However, -undefined dynamic_lookup is already deprecated

@kateinoigakukun Where did you find this out please? I'm struggling to find useful documentation about this option for the darwin linker. The `ld` manpage on my Macbook simply says

> Note: dynamic_lookup that depends on lazy binding will not work with chained fixups

**Comment by kateinoigakukun** [comment]
> The ld manpage on my Macbook simply says

Actually, the option is explicitly deprecated on some platforms and effectively deprecated on macOS too (e.g. no chained-fixup support): https://github.com/apple-opensource/ld64/blob/e28c028b20af187a16a7161d89e91868a450cadc/src/ld/Options.cpp#L5716

The `dynamic_lookup` mechanism prevents several optimizations, so I don't think the support will be continued.

**Comment by eightbitraptor** [comment]
Closing this in favour of https://github.com/ruby/ruby/pull/10456

**Pull request closed without merging**